keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,604 @@
+
+### CSI API
+The CSI create request will be extended to add provisioned IO parameters. For volume creation, cloud providers can add iops and throughput field in parameters and process in the csi driver, an example:
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gcepd
+provisioner: pd.csi.storage.gke.io
+parameters:
+  type: hyperdisk-balanced
+  provisioned-iops-on-create: '10000'
+  provisioned-throughput-on-create: '1000Mi'
+volumeBindingMode: WaitForFirstConsumer
+```
+
+A new ControllerModifyVolume RPC will be added. The current spec requires ["If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK"](https://github.com/container-storage-interface/spec/blob/master/spec.md). The change here will be if the capacity change is for iops and throughput, we will not require that to be larger than the current configuration.
+
+#### ControllerModifyVolume
+
+A Controller plugin SHALL implement this RPC call . This RPC allows the CO to change iops and throughput of a volume.
+
+This operation MUST be idempotent. If a volume corresponding to the specified volume ID is already equal to the target iops/throughput of the update request, the plugin SHOULD reply 0 OK.
+
+```
+// ControllerServer is the server API for Controller service.
+type ControllerServer interface {
+        ...
+        ControllerModifyVolume(context.Context, *ControllerModifyVolumeRequest) (*ControllerModifyVolumeResponse, error)
+        ...
+}
+
+type ControllerModifyVolumeRequest struct {
+        // The ID of the volume to expand. This field is REQUIRED.
+        VolumeId string
+        Iops string 
+        Throughput string
+}
+type ControllerModifyVolumeResponse struct {
+        Iops int64
+        Throughput int64
+}
+```
+
+For future first class QoS parameters, the API fields can be expanded.
+
+The according errors:
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Exceeds capabilities | 3 INVALID_ARGUMENT | Indicates that the CO has specified capabilities not supported by the volume. | Caller MAY verify volume capabilities by calling ValidateVolumeCapabilities and retry with matching capabilities. |
+| Volume does not exist | 5 NOT FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
+| Volume in use | 9 FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be updated because it is currently published on a node but the plugin does not have ONLINE expansion capability. | Caller SHOULD ensure that volume is not published and retry with exponential back off. |
```

Comment:
I guess this hints that there would need to be another capability stating whether volume modify will work for an ONLINE volume

Reply:
We want to use capabilities in the CSI layer but we do not want to add a new flag to StorageClass.